### DISCUSSION TITLE
Developer tools for debugging

### DISCUSSION THREAD

**Discussion by fbwoolf** [blockstack/stacks-wallet-web#1341] [General]
What are the best tools people are using to debug our code? I don't have access to class names and/or components in the React dev tools. What am I missing here?

These images are taken from the explorer, but starting the discussion here bc I believe it is the same in the web wallet and would love help understanding the best dev tools to work with our UI library.

![image](https://user-images.githubusercontent.com/6493321/123146335-5c27fb00-d423-11eb-8c8b-8d0abddeaaf6.png)

![image](https://user-images.githubusercontent.com/6493321/123146342-5f22eb80-d423-11eb-939d-e5da465d5376.png)

**Comment by aulneau** [comment]
These issues are not exactly due to our UI library, but our general practices (or lack thereof) when writing components.

For the CSS classnames, we might need to add the emotion babel plugin for better class names during dev in the UI library, we have it already in the explorer. https://emotion.sh/docs/@emotion/babel-plugin

For the `Anonymous` components in the react dev tools, this is due to how we make use of forwardRef or memo, and how the internal structure of the UI components are built. You can add filters to the dev tools to filter only the things you care about:

<img width="1511" alt="Screen Shot 2021-06-23 at 1 53 02 PM" src="https://user-images.githubusercontent.com/11803153/123152687-744f4880-d42a-11eb-81cc-a4e2e451d611.png">

<img width="1589" alt="Screen Shot 2021-06-23 at 1 54 04 PM" src="https://user-images.githubusercontent.com/11803153/123152738-84672800-d42a-11eb-853d-f6e878b82353.png">

**Reply by fbwoolf** [reply]
So the babel plugin should be doing this: https://sheelahb.com/blog/include-emotion-component-names-in-css-classes ...it should be attaching the component name to the css class name. I don't think it is working in the explorer either, bc the image I took is from the explorer. I'll look into how it is setup.

**Reply by fbwoolf** [reply]
Thanks for the tip on filtering, I'll try that. Sometimes I don't understand why there is so much code to do such simple things. What are all these components I'm filtering out?

**Reply by aulneau** [reply]
> So the babel plugin should be doing this: https://sheelahb.com/blog/include-emotion-component-names-in-css-classes ...it should be attaching the component name to the css class name. I don't think it is working in the explorer either, bc the image I took is from the explorer. I'll look into how it is setup.

Yeah! for sure -- it could be that we need to make a modification in the UI library to better support dev mode. 

May I ask, what is it that you're hoping for as far as a dev experience goes for needing CSS class names? I personally have not thought of needing a className in years, and it might be because I'm thinking about the components/styles differently.

> Thanks for the tip on filtering, I'll try that. Sometimes I don't understand why there is so much code to do such simple things. What are all these components I'm filtering out?

Of course! A lot of the components are related to if they are wrapped in `forwardRef` or `memo`. You can read about both: 
- https://reactjs.org/docs/forwarding-refs.html
- https://reactjs.org/docs/react-api.html#reactmemo

But the high level is that all of the components from `@stacks/ui` have `forwardRefWithAs` which is a wrapper on top of `React.forwardRef` that is used for two reasons: all the components need to be able to accept a `ref`, and they all can accept an `as` prop. The `as` prop will change the underlying component or DOM element. Take this as an example:

```tsx
// the underlying DOM element will be a `<label>`
const MyLabel = () => <Box as="label" />

// and it will by default automatically accept 
// all the types associated with the HTML label element, 
// so you can do this:
const MyLabel = () => <Box as="label" htmlFor="some-id" /> // htmlFor is a `<label>` only attribute
// and there won't be any runtime TypeScript errors

// you can also do this:
const CaptionLabel = () => <Box as={MyLabel} color="whitesmoke" />
// which will render as the `<label>` component but also have the css of `color: whitesmoke;`
```

**Reply by aulneau** [reply]
The concept of the `as` prop is sometimes referred to as `polymorphic`, here are some other examples / context:
https://github.com/kripod/react-polymorphic-box
https://radix-ui.com/primitives/docs/utilities/polymorphic
https://react-typescript-cheatsheet.netlify.app/docs/advanced/patterns_by_usecase/#polymorphic-components-eg-with-as-props
https://labs.walltowall.com/calico-polymorphic-components/

**Reply by aulneau** [reply]
> Sometimes I don't understand why there is so much code to do such simple things.

I wonder about this too, but then I remember that even though it seems like they are "simple things", they are all actually pretty complex! and it's nice that all the complexity is hidden away and we can just do something like this:

```tsx
const MyCoolButton = (props: BoxProps) => <Box as="button" size={[null, 'sm', 'lg']} display={['none', 'block']} color={[null, 'blue', 'red']} />
```
This will give us a `<button>` that will not be visible on mobile breakpoints, on tablet/mid sized screens the button will be small and the color will be blue, and on large screens button will be large and the color will be red!

**Reply by fbwoolf** [reply]
> May I ask, what is it that you're hoping for as far as a dev experience goes for needing CSS class names? I personally have not thought of needing a className in years, and it might be because I'm thinking about the components/styles differently.
> 

I typically look for class names or component names in the UI and then match them up in the code to know where I need to make changes. As someone new to the code base, I need a reference point. I just came from working with a pretty robust design system built purely in css, so we relied heavily on classNames ...not CSS in JS.

Definitely good component names and the ability to use the React dev tools replaces needing classNames, I agree.

**Reply by aulneau** [reply]
> I typically look for class names or component names in the UI and then match them up in the code to know where I need to make changes. As someone new to the code base, I need a reference point. I just came from working with a pretty robust design system built purely in css, so we relied heavily on classNames ...not CSS in JS.

> Definitely good component names and the ability to use the React dev tools replaces needing classNames, I agree.

Yeah, that makes total sense! coming from a world where more vanilla CSS is used, this would be a big jump and shift towards something different.

For me, and maybe this will help you, I typically rely pretty heavily on the select element tool in both the standard chrome dev tools and the react dev tools panel, which then helps me see the component tree a bit better and find the component I'm looking for.

https://user-images.githubusercontent.com/11803153/123165669-23475080-d43a-11eb-8eea-19806d618641.mp4

I think using practices like these can really help with debugging, and I agree, I def think we should dig into how better to add component names to classes in dev mode if possible!

**Reply by fbwoolf** [reply]
Yep, that is what I use! 💯     Btw, thanks for the links!

**Reply by fbwoolf** [reply]
@aulneau filtering components is helping a lot! 👍
